ako-operator/controllers/statefulset.go

```diff
@@ -0,0 +1,253 @@
+		return v1.PullNever, nil
+	default:
+		return v1.PullPolicy(""), errors.New("invalid pull policy")
+	}
+}
+
+func buildResources(ako akov1alpha1.AKOConfig) (v1.ResourceRequirements, error) {
+	var rr v1.ResourceRequirements
+
+	limitCPU, err := resource.ParseQuantity(ako.Spec.Resources.Limits.CPU)
+	if err != nil {
+		return rr, err
+	}
+	limitMemory, err := resource.ParseQuantity(ako.Spec.Resources.Limits.Memory)
+	if err != nil {
+		return rr, err
+	}
+
+	requestedCPU, err := resource.ParseQuantity(ako.Spec.Resources.Requests.CPU)
+	if err != nil {
+		return rr, err
+	}
+
+	requestedMemory, err := resource.ParseQuantity(ako.Spec.Resources.Requests.Memory)
+	if err != nil {
+		return rr, err
+	}
+
+	return v1.ResourceRequirements{
+		Limits: v1.ResourceList{
+			"cpu":    limitCPU,
+			"memory": limitMemory,
+		},
+		Requests: v1.ResourceList{
+			"cpu":    requestedCPU,
+			"memory": requestedMemory,
+		},
+	}, nil
+}
+
+func BuildStatefulSet(ako akov1alpha1.AKOConfig) (appsv1.StatefulSet, error) {
+	sf := appsv1.StatefulSet{}
+
+	sf.ObjectMeta = metav1.ObjectMeta{
+		Name:      StatefulSetName,
+		Namespace: AviSystemNS,
+	}
+
+	image := ako.Spec.ImageRepository
+	imagePullPolicy, err := getPullPolicy(ako.Spec.ImagePullPolicy)
+	if err != nil {
+		return sf, err
+	}
+	var replicas int32 = 1
+	sf.Spec.Replicas = &replicas
+	sf.Spec.ServiceName = ServiceName
+	akoLabels := map[string]string{
+		"app": "ako",
+	}
+	sf.Spec.Selector = &metav1.LabelSelector{
+		MatchLabels: akoLabels,
+	}
+
+	// build the env vars
+	envVars := getEnvVars(ako)
+
+	volumeMounts := []v1.VolumeMount{}
+	volumes := []v1.Volume{}
+	if ako.Spec.PersistentVolumeClaim != "" {
+		volumeMounts = append(volumeMounts, v1.VolumeMount{
+			Name:      "ako-pv-storage",
+			MountPath: ako.Spec.MountPath,
+		})
+		volumes = append(volumes, v1.Volume{
+			Name: "ako-pv-storage",
+			VolumeSource: v1.VolumeSource{
+				PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+					ClaimName: ako.Spec.PersistentVolumeClaim,
+				},
+			},
+		})
+	}
+
+	apiServerPort := ako.Spec.APIServerPort
+	if apiServerPort == 0 {
+		apiServerPort = 8080
+	}
+
+	resources, err := buildResources(ako)
+	if err != nil {
+		return sf, err
+	}
+	template := v1.PodTemplateSpec{}
+	template.SetLabels(akoLabels)
+	template.Spec = v1.PodSpec{
+		ServiceAccountName: ServiceAccountName,
+		Volumes:            volumes,
+		Containers: []v1.Container{
+			{
+				Name:            "ako",
+				VolumeMounts:    volumeMounts,
+				Image:           image,
+				ImagePullPolicy: imagePullPolicy,
+				Lifecycle: &v1.Lifecycle{
+					PreStop: &v1.Handler{
+						Exec: &v1.ExecAction{
+							Command: []string{"/bin/sh", "/var/pre_stop_hook.sh"},
+						},
+					},
+				},
+				Ports: []v1.ContainerPort{
+					{
+						Name:          "http",
+						ContainerPort: 80,
+						Protocol:      "TCP",
+					},
+				},
+				Resources: resources,
+				LivenessProbe: &v1.Probe{
+					InitialDelaySeconds: 5,
```

Comment:
How does someone change these values with the operator mode?

Reply:
As discussed, they can change it in the statefulset directly and the values will be retained till a user makes a change in the `AKOConfig` properties forcing the operator to restart the statefulset.